mlvm/mlvm.sh

```diff
@@ -0,0 +1,197 @@
+
+set -euxo pipefail
+
+readonly JARS_DIR=/usr/lib/spark/jars
+readonly CONNECTORS_DIR=/usr/local/share/google/dataproc/lib
+
+# readonly DEFAULT_INIT_ACTIONS_BUCKET=gs://dataproc-initialization-actions
+# readonly INIT_ACTIONS_BUCKET="$(/usr/share/google/get_metadata_value attributes/INIT_ACTIONS_BUCKET ||
+#   echo ${DEFAULT_INIT_ACTIONS_BUCKET})"
+readonly DEFAULT_INIT_ACTIONS_REPO="https://github.com/GoogleCloudDataproc/initialization-actions.git"
+readonly INIT_ACTIONS_REPO="$(/usr/share/google/get_metadata_value attributes/INIT_ACTIONS_REP ||
+  echo ${DEFAULT_INIT_ACTIONS_REPO})"
+readonly INIT_ACTIONS_BRANCH="$(/usr/share/google/get_metadata_value attributes/INIT_ACTIONS_BRANCH ||
+  echo 'master')"
+readonly INIT_ACTIONS_DIR=/opt/init-actions
+readonly include_gpus="$(/usr/share/google/get_metadata_value attributes/include-gpus || true)"
+
+# Conda
+BASE_R_PACKAGES=(
+  "r-essentials=3.6.0"
+  "r-xgboost=0.90.0.2"
+  "r-sparklyr=1.0.0"
+)
+
+# Pip
+BASE_PYTHON_PACKAGES=(
+  "google-cloud-bigquery==1.26.1" 
+  "google-cloud-datalabeling==0.4.0"
+  "google-cloud-storage==1.30.0"
+  "google-cloud-bigtable==1.4.0" 
+  "google-cloud-dataproc==1.0.1" 
+  "google-api-python-client==1.10.0" 
+  "matplotlib==3.3.0"
+  "mxnet==1.6.0" 
+  "nltk==3.5"
+  "numpy==1.18.4" 
+  "rpy2==3.3.3"
+  "scikit-learn==0.23.1" 
+  "sparksql-magic==0.0.3" 
+  "tensorflow-datasets==3.2.1"
+  "tensorflow-estimator==2.3.0"
+  "tensorflow-hub==0.8.0"
+  "tensorflow-io==0.14.0"
+  "tensorflow-probability==0.10.1" 
+  "torch==1.5.1" 
+  "torchvision==0.6.1" 
+  "xgboost==1.1.0"
+)
+
+if [[ -n ${include_gpus} ]]; then
+  BASE_PYTHON_PACKAGES+=("tensorflow-gpu==2.2.0")
+else
+  BASE_PYTHON_PACKAGES+=("tensorflow==2.2.0")
+fi
+
+
+if [ "$(echo "$DATAPROC_VERSION >= 2.0" | bc)" -eq 1 ]; then 
+  BASE_PYTHON_PACKAGES+=("spark-tensorflow-distributor==0.1.0")
+fi
+
+mkdir -p ${JARS_DIR}
+mkdir -p ${CONNECTORS_DIR}
+mkdir -p ${INIT_ACTIONS_DIR}
+
+function execute_with_retries() {
+  local -r cmd=$1
+  for ((i = 0; i < 10; i++)); do
+    if eval "$cmd"; then
+      return 0
+    fi
+    sleep 5
+  done
+  echo "Cmd \"${cmd}\" failed."
+  return 1
+}
+
+function download_spark_jar() {
+  local -r url=$1
+
+  wget -nv --timeout=30 --tries=5 --retry-connrefused \
+  -P "${JARS_DIR}" "${url}" 
+
+}
+
+function download_init_actions() {
+  # Download initialization actions locally.
+  git clone -b "${INIT_ACTIONS_BRANCH}" --single-branch "${INIT_ACTIONS_REPO}" "${INIT_ACTIONS_DIR}"
+  find "${INIT_ACTIONS_DIR}" -name '*.sh' -exec chmod +x {} \;
+}
+
+function install_connectors() {
+  "${INIT_ACTIONS_DIR}/connectors/connectors.sh"
+}
+
+function install_python_packages() {
+  readonly EXTRA_PYTHON_PACKAGES="$(/usr/share/google/get_metadata_value attributes/PYTHON_PACKAGES || true)"
+
+  # Installing all at once is flaky. Added this here. 
+  for package in "${BASE_PYTHON_PACKAGES[@]}"; do
+    execute_with_retries "pip install $package"
```

Comment:
Pass all packages in a single command: `pip install "${BASE_PYTHON_PACKAGES[@]}"`

Reply:
I originally had this but it would flake by occasionally leading to some closed connections from PyPi. I'll try again with the wrapper retry function.